In Likes and Comment Likes modules, prevent blocking page load with 2 scripts

kienstra opened the pull request:
Fixes #13929


#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* If likes are disabled globally or for a post, prevents enqueueing scripts for the Likes or Comment Likes modules
* Passes to `wp_enqueue_script()` the 5th `$in_footer` argument of `true`, to prevent the scripts from blocking the page load
* [These](https://github.com/Automattic/jetpack/blob/8d04b28cff9294608111d3293fb2b31a5fba1191/_inc/postmessage.js) [scripts](https://github.com/Automattic/jetpack/blob/8d04b28cff9294608111d3293fb2b31a5fba1191/_inc/jquery.jetpack-resize.js) look to be mainly jQuery plugins
* So far, the 'Like' button seems to work fine with these in the footer. But this could of course use more testing.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* It modifies the performance of an existing feature, no functional change intended.

#### Testing instructions:
1. On a non-local site, ensure 'Add Like buttons...' is toggled on:
![sharing-settings](https://user-images.githubusercontent.com/4063887/75521644-c797bf00-59cd-11ea-9257-fe25a5a2dd69.png)
2. Go to the front-end of a post
3. Expected: The 'Like' button should appear, and should be clickable:

![like-button](https://user-images.githubusercontent.com/4063887/75521700-e5fdba80-59cd-11ea-9671-8283a9786a60.png)

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Performance enhancement for the Likes module

kienstra committed:
Prevent blocking page load with 2 scripts in Like module

These both depend on jquery,
so this could remove even another script
from blocking rendering.

kienstra commented on modules/likes.php:
```diff
@@ -280,8 +280,8 @@ class_exists( 'Jetpack_AMP_Support' )
-			wp_enqueue_script( 'jetpack_resize', '/wp-content/js/jquery/jquery.jetpack-resize.js', array( 'jquery' ), JETPACK__VERSION, false );
+			wp_enqueue_script( 'postmessage', '/wp-content/js/postmessage.js', array( 'jquery' ), JETPACK__VERSION, true );
+			wp_enqueue_script( 'jetpack_resize', '/wp-content/js/jquery/jquery.jetpack-resize.js', array( 'jquery' ), JETPACK__VERSION, true );
 			wp_enqueue_script( 'jetpack_likes_queuehandler', plugins_url( 'queuehandler.js' , __FILE__ ), array( 'jquery', 'postmessage', 'jetpack_resize' ), JETPACK__VERSION, true );
```
This [script](https://github.com/Automattic/jetpack/blob/8d04b28cff9294608111d3293fb2b31a5fba1191/modules/likes/queuehandler.js) that depends on 2 scripts above is already in the footer, with the `$in_footer` argument  of`true`

kienstra commented:
**Request For Review**

Hi @dero,
Really good idea for the [Map Block](https://github.com/Automattic/jetpack/issues/14827) earlier.

Could you please review this when you have a chance? 

Thanks, Dero!

dero reviewed (commented):
@kienstra The changes look good, I've verified the scripts get loaded before the `https://widgets.wp.com/likes/master.html` iframe (which depends of them being loaded) in the footer.

I'd suggest we make the same change in the sibling `comment-likes.php` module as well. These two seem the only ones to pull those scripts in.

kienstra commented:
Nice, thanks for reviewing this.

>I'd suggest we make the same change in the sibling comment-likes.php module as well. These two seem the only ones to pull those scripts in.

Oh, good idea.

kienstra committed:
Don't enqueue script if the Likes module isn't visible

As Dan raised,
this should not enqueue this JS if this
isn't in the document.

kienstra commented:
https://github.com/Automattic/jetpack/pull/14841/commits/716e19ddea4d02cb0f9309de9cd00c2f98dedae4 addresses #13929 for the Likes module, not Comment Likes.

It prevents the JS from being enqueued when Likes are disabled for a post or globally.

Steps To Test
1. In a non-local site, ensure 'Add like buttons' is toggled on:
![sharing-settings](https://user-images.githubusercontent.com/4063887/75521644-c797bf00-59cd-11ea-9257-fe25a5a2dd69.png)

2. On a post, uncheck 'Show likes':
<img width="844" alt="show-likes-check" src="https://user-images.githubusercontent.com/4063887/75825773-adfacc80-5d6b-11ea-85d8-7b3f72f8285b.png">

3. Go to the front-end permalink, not just the preview URL
4. Search the source for `queuehandler.js`
5. Expected: it's not in the source:
<img width="789" alt="not-found" src="https://user-images.githubusercontent.com/4063887/75825641-607e5f80-5d6b-11ea-85cb-cb14ebc23bf2.png">
6. Toggle off 'Add like buttons' from step 1 above 
7. Repeat steps 3-5, the script should still not be in the source

kienstra committed:
[not verified] Apply a similar change to the Comment Likes module

Don't enqueue scripts if likes aren't enabled,
and move the dependencies to the footer,
as the script that depends on them,
'jetpack_likes_queuehandler',
is in the footer.
It doesn't seem necessary for these to be above
the footer.

kienstra committed:
Fix a unit test that asserted the opposite

It should ensure that the styles are present,
not that they aren't.

kienstra commented:
**Request For Review**

Hi @dero,
Thanks your [suggestion](https://github.com/Automattic/jetpack/pull/14841#pullrequestreview-368125199):

>I'd suggest we make the same change in the sibling comment-likes.php module as well. These two seem the only ones to pull those scripts in.

Could you please review this again, now that it addresses your suggestion?

It also [addresses issue #13929](https://github.com/Automattic/jetpack/pull/14841#issuecomment-594203358)

Thanks, Dero!

josephscott commented:
> Passes to wp_enqueue_script() the 5th $in_footer argument of true, to prevent the scripts from blocking the page load

Has it been confirmed that this change of moving the script to the footer impacts how quickly the page renders?

I ask because in general moving a script tag to the bottom of the page will not change the impact in terms of render blocking.  Browsers like Chrome will quickly scan the HTML looking for resources it thinks will be important to rendering the page and loads them early on, mainly CSS and JS.  Even if they are listed at the very bottom of the page.

You can see that happening with this test - https://www.webpagetest.org/result/200304_1E_35a21961f2be5996d4a1613a687f7c3b/1/details/#waterfall_view_step1 - the third request in the page load was from a script tag at the bottom of the page ( bundle.js ).  Specifically line 1333, out of 1436 lines.

kienstra commented:
Hi @josephscott,
Thanks for testing that, and sharing the results of the [webpagetest](https://www.webpagetest.org/result/200304_1E_35a21961f2be5996d4a1613a687f7c3b/1/details/#waterfall_view_step1).

No, I haven't tested the performance difference between [passing](https://developer.wordpress.org/reference/functions/wp_enqueue_script/#parameters) the 5th `$in_footer` argument as `true`, and not passing it. 

That's a general best practice, but you're probably right that it's not a big impact.

I'll look at it in more detail.

kienstra commented:
## Performance impact of moving scripts to footer

Hi @josephscott,
Sorry for the delay.

Like you mentioned, the performance benefit of only moving the scripts to the footer looks small.

In 10 trials, the average time to First Contentful Paint was 3.15 with the scripts moved to the footer, as opposed to 3.4 with them above the footer. 

But most of that difference is from outliers. The most common case is that there is no difference.

## Steps to reproduce
1. `checkout` this PR's branch
2. Ensure 'Add like buttons' is toggled on:
![sharing-settings](https://user-images.githubusercontent.com/4063887/75521644-c797bf00-59cd-11ea-9257-fe25a5a2dd69.png)

3. Activate [Twenty Twenty](https://wordpress.org/themes/twentytwenty/)
4. Create a simple post, and publish it
5. In Chrome, go to the front-end of it on an Incognito window 
6. Run 10 Lighthouse audits for Mobile
7. Revert this PR's diff for the 5th argument of `wp_enqueue_script()`:

```diff
diff --git a/modules/comment-likes.php b/modules/comment-likes.php
index 8868f2688..4332e8eb5 100644
--- a/modules/comment-likes.php
+++ b/modules/comment-likes.php
@@ -142,7 +142,7 @@ class Jetpack_Comment_Likes {
                        Assets::get_file_url_for_environment( '_inc/build/postmessage.min.js', '_inc/postmessage.js' ),
                        array( 'jquery' ),
                        JETPACK__VERSION,
-                       true
+                       false
                );
                wp_enqueue_script(
                        'jetpack_resize',
@@ -152,7 +152,7 @@ class Jetpack_Comment_Likes {
                        ),
                        array( 'jquery' ),
                        JETPACK__VERSION,
-                       true
+                       false
                );
                wp_enqueue_script( 'jetpack_likes_queuehandler', plugins_url( 'likes/queuehandler.js' , __FILE__ ), array( 'jquery', 'postmessage', 'jetpack_resize' ), JETPACK__VERSION, true );
        }
diff --git a/modules/likes.php b/modules/likes.php
index 4a0b0273a..f4d3f683d 100644
--- a/modules/likes.php
+++ b/modules/likes.php
@@ -296,7 +296,7 @@ class Jetpack_Likes {
                        Assets::get_file_url_for_environment( '_inc/build/postmessage.min.js', '_inc/postmessage.js' ),
                        array( 'jquery' ),
                        JETPACK__VERSION,
-                       true
+                       false
                );
                wp_register_script(
                        'jetpack_resize',
@@ -306,7 +306,7 @@ class Jetpack_Likes {
                        ),
                        array( 'jquery' ),
                        JETPACK__VERSION,
-                       true
+                       false
                );
                wp_register_script(
                        'jetpack_likes_queuehandler',

```
8. Run another 10 Lighthouse audits for Mobile

## Results

First Contentful Paint in seconds
Without/with this PR's moving of scripts to footer 

| Without this PR | With this PR |
|---|---|
| 3.8  | 3.3  |
| 3.3  | 3.3  |
| 3.3  | 3.3  |
| 3.3 | 3.3 |
| 3.8 | 3.3 |
| 3.3 | 3.3 |
| 3.3 | 2.2 |
| 3.3 | 2.9 |
| 3.3 | 3.3 |
| 3.3 | 3.3 |

### Average
| Without this PR | With this PR |
|---|---|
| 3.4 | 3.15 |

### Difference
0.25 seconds, on average
Still, the most common case is that there's no difference

dero commented:
> I ask because in general moving a script tag to the bottom of the page will not change the impact in terms of render blocking.

@josephscott In my understanding [speculative parsing](https://developer.mozilla.org/en-US/docs/Glossary/speculative_parsing) is only a heuristic that is supposed to "guess" which assets will be needed for page construction and thus it only affects the _timing_ of network requests. The synchronous script _execution_ still blocks the HTML parser and thus introduces delays in page render.

To verify this, I've inspected the main thread tasks sequence on `master` and this branch and checked which segments of the incoming HTML markup are parsed when.

**On `master`:**

* **0 ms mark**: [first 260 lines](https://user-images.githubusercontent.com/433570/75978808-9cd2cc80-5edf-11ea-80dd-a5d85f629574.png) of markup are parsed.
* Then `style.css` is parsed synchronously.
* Then all synchronous scripts in `<head>` are executed up until `jquery.jetpack-resize.js`.
* **725 ms mark**: [lines 264..679](https://user-images.githubusercontent.com/433570/75978974-e3282b80-5edf-11ea-8a66-8d4290e8df41.png) of markup are parsed.
* Then all synchronous scripts from the footer of the page are executed.
* **1025 ms mark**: the rest of the markup is parsed.

**On this branch:**

* **0 ms mark**: [first 260 lines](https://user-images.githubusercontent.com/433570/75980158-39966980-5ee2-11ea-88d3-e5451169262c.png) of markup are parsed.
* Then `style.css` is parsed synchronously.
* Then all synchronous scripts in `<head>` are executed up until `jquery.migrate.js`.
* **590 ms mark**: [lines 262..677](https://user-images.githubusercontent.com/433570/75979920-cd1b6a80-5ee1-11ea-83c9-a7c357205a03.png) of markup are parsed.
  * Notice on screenshot: This happens while `postmessage.js` and `jquery.jetpack-resize.js` are still being loaded.
* Then all synchronous scripts from the footer of the page are executed including the ones being moved by this PR.
* **880 ms mark**: the rest of the markup is parsed.

**Based on this experiment, I'd say we can conclude that:**

* Speculative parsing allows assets to start loading even before the parser processes their `<script>` element.
* The engine still halts the renderer when it encounters a synchronous style / script, which means that sequencing matters.
* On a simple page with simple a small blocking assets, the gains may be imperceptible, but the principle still holds.

cc @kienstra

josephscott commented:
@dero now this is the what I like!  When a claim is made about performance, it needs to be confirmed.

kienstra commented:
Yeah, thanks a lot for your really detailed look at this! 😄

kraftbj commented:
r203942-wpcom
